src/polygon/pickle-jar-univ3.sol

```diff
@@ -264,6 +224,155 @@ contract PickleJarUniV3Poly is ERC20, ReentrancyGuard {
+            (token1Amount != 0)
+        ) token1.safeTransferFrom(msg.sender, address(this), token1Amount);
+
+        _balanceProportion(getLowerTick(), getUpperTick());
+    }
+
+    function _convertMatic(uint256 token0Amount, uint256 token1Amount)
+        internal
+        returns (uint256, uint256)
+    {
+        uint256 _matic = address(this).balance;
+        if (_matic > 0) {
+            WETH(wmatic).deposit{value: _matic}();
+
+            if (address(token0) == wmatic) {
+                token0Amount = _matic;
+            } else if (address(token1) == wmatic) {
+                token1Amount = _matic;
+            }
+        }
+        return (token0Amount, token1Amount);
+    }
+
+    function _refundMatic(uint256 tokenAmount) internal {
+        uint256 _refund = WETH(wmatic).balanceOf(address(this)).sub(
+            tokenAmount
+        );
+        WETH(wmatic).withdraw(_refund);
+        (bool sent, bytes memory data) = (msg.sender).call{value: _refund}("");
+        require(sent, "Failed to refund Matic");
+    }
+
+    function _refundUnused() internal returns (uint256) {
+        PoolVariables.Info memory _cache;
+        _cache.amount0Desired = token0.balanceOf(address(this));
+        _cache.amount1Desired = token1.balanceOf(address(this));
+
+        _cache.liquidity = (
+            pool.liquidityForAmounts(
+                _cache.amount0Desired,
+                _cache.amount1Desired,
+                getLowerTick(),
+                getUpperTick()
+            )
+        );
+
+        (_cache.amount0, _cache.amount1) = pool.amountsForLiquidity(
+            _cache.liquidity,
+            getLowerTick(),
+            getUpperTick()
+        );
+
+        if (_cache.amount0Desired > _cache.amount0)
+            if (address(token0) == address(wmatic))
+                _refundMatic(_cache.amount0Desired.sub(_cache.amount0));
+            else {
+                token0.safeTransfer(
+                    msg.sender,
+                    _cache.amount0Desired.sub(_cache.amount0)
+                );
+            }
+
+        if (_cache.amount1Desired > _cache.amount1)
+            if (address(token1) == address(wmatic))
+                _refundMatic(_cache.amount1Desired.sub(_cache.amount1));
+            else {
+                token1.safeTransfer(
+                    msg.sender,
+                    _cache.amount1Desired.sub(_cache.amount1)
+                );
+            }
+        return _cache.liquidity;
+    }
+
+    function _balanceProportion(int24 _tickLower, int24 _tickUpper) internal {
+        PoolVariables.Info memory _cache;
+
+        _cache.amount0Desired = token0.balanceOf(address(this));
+        _cache.amount1Desired = token1.balanceOf(address(this));
+
+        (uint160 sqrtPriceX96, , , , , , ) = pool.slot0();
+        uint160 sqrtRatioAX96 = TickMath.getSqrtRatioAtTick(_tickLower);
+        uint160 sqrtRatioBX96 = TickMath.getSqrtRatioAtTick(_tickUpper);
+
+        _cache.liquidity = uint128(
```

Comment:
Fascinating! Is this the secret sauce for getting the achievable liquidity with a single token input?